src/Controller/Component/AuthComponent.php

```diff
@@ -391,6 +392,27 @@ protected function _unauthenticated(Controller $controller)
     }
 
     /**
+     * Returns the URL of the login action to redirect to.
+     *
+     * This includes the redirect query string if applicable.
+     *
+     * @return array|string
+     */
+    protected function _loginActionRedirectUrl()
+    {
+        $currentUrl = $this->request->here(false);
+
+        $loginAction = $this->_config['loginAction'];
+        if (is_array($loginAction)) {
+            $loginAction['?'][static::QUERY_STRING_REDIRECT] = $currentUrl;
+        } else {
+            $loginAction .= '?' . static::QUERY_STRING_REDIRECT . '=' . rawurlencode($currentUrl);
```

Comment:
Anyone setting login action as string with query string deserves to face problems :P